addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/rekognition/ImageHelper.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * (C) Copyright 2018 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Gethin James
+ */
+package org.nuxeo.ai.rekognition;
+
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ai.enrichment.EnrichmentUtils;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.blob.ManagedBlob;
+import org.nuxeo.ecm.core.storage.sql.ImageHelperWithS3;
+import com.amazonaws.services.rekognition.model.Image;
+import com.amazonaws.services.rekognition.model.S3Object;
+
+/**
+ * Helps with S3 images
+ */
+public class ImageHelper {
+
+    protected static final ImageHelper INSTANCE = new ImageHelper();
+
+    private static final Log log = LogFactory.getLog(ImageHelper.class);
+
+    protected final boolean hasS3Storage;
+
+    protected ImageHelperWithS3 s3Helper;
+
+    public ImageHelper() {
+
+        boolean hasS3BinaryManager = true;
+        try {
+            Class.forName("org.nuxeo.ecm.core.storage.sql.S3BinaryManager");
```

Comment:
Its's a common pattern in frameworks like Spring to have a static initiallizer to check for classes.  I get your point that "may not be used for a given blob" so will rework it to check for the String name every time.

Reply:
There are 3 scenarios that I want to cover when deciding whether to pass on S3 object reference or encode the bytes:

1. The `S3BinaryManager` isn't in the classpath.
2. The `S3BinaryManager` is available but it may not be the given blobprovider.
3. The `S3BinaryManager` is available but the user has chosen to never pass a reference - always the blob bytes.

**Current implementation**
For (1), the code only instantiates the class (`ImageHelperWithS3`) that uses the `S3BinaryManager` if its available in the classpath.  I feel this is clear and safe.
For (2), I would return `null`, so need to fix that.
For (3), that's covered in the `getInstance` method.

It sounds like your proposal is to always instantiate `ImageHelperWithS3`, this would be creating an object with methods on it that can't be used.  Off the top of my head, I don't know if the class loader issues would happen on instantiation or when the methods are used; propably the latter.  To me its clearer to only instantiate `ImageHelperWithS3` if S3BinaryManager` is available in the classpath.

Reply:
No, I'm suggesting to have a generic `ImageHelper` whose method check the class name of the blob provider (`blobProvider.getClass().getName().equals("...S3BinaryManager")`) and if that's the case pass it to the `ImageHelperWithS3.getImage` method which can then have code specific to the class. There should be no static reference to S3 in the base `ImageHelper` class.